Change DBUS service to session bus

ndim opened the pull request:
Change the soundcraft-utils DBUS service from the system bus
to the session bus. This means

  * changes to the dbus service and client code
    (use `SessionBus()` instead of `SystemBus()`)

  * changes to the dbus config files (no policy file required,
    no `User=root`, different location for *.service file)

  * changes to the dbus config file install location
    in `soundcraft_dbus_service --setup`:

    If service executable is
    `(/usr/local|/usr|~/.local)/(bin|libexec|sbin)/soundcraft_dbus_service`
    use
    `(/usr/local|/usr|~/.local)/share/dbus-1`.

  * change program messages

TODO:

  * Update `README.md` section on (un)install, added XXXXXX markers
    as reminders.

ndim committed:
Add `--version` argument to all installed scripts

Add a `--version` argument to all installed scripts which prints
a basic version information message for good form.

This also establishes how to add a command line argument to the
Gtk app. That might come in useful later, e.g. when adding a
`--no-dbus` option for the Gtk app.

Note that if you wanted to have a more complex multiline version
information message like e.g. the GNU tools do, you cannot use
the `argparse` `action="version"` because argparse internally
interprets newlines in the `version=` message just like any other
whitespace character.

lack committed:
Re-run pre-commit checks on every pull request to 'release'

lack committed:
Merge pull request #20 from ndim/add-cmdline-version-argument

Add `--version` argument to all installed scripts

ndim committed:
Use `dbus` strings instead of docstrings for pydbus XML

Use `dbus` string members instead of docstrings to define the pydbus
interface description XML.

As `pydbus` will happily use a `dbus` member for the interface description
XML, this frees the docstring giving us the opportunity to write an actual
docstring.

ndim committed:
Align DBus XML interface descriptions

This improves human readability of the DBus interface definition XML.

ndim committed:
Use a list of classes instead of eval()ing a string

Directly iterate over a list of classes instead of over a
list of strings which were then used to construct a string
to be handed to `eval()`.

ndim committed:
Correctly format firmware version BCD number as string

Correctly format the firmware version BCD number as string,
such as e.g. "1.09".

This does not use an f string, as f"{major}.{minor:02}" makes
the pre-commit checks complain about invalid formatting.

ndim commented:
Better not merge this yet. There is still a lot to do.

I have about a dozen more commits locally which contain parts which should be committed eventually, but not necessarily in their current state and sequence.

Do you want me to occasionally force-push that dozen here, the commit messages starting with `XXX` to mark them very obviously as work in progress not suitable yet for merging?

lack commented:
I'm fine with you force-pushing here any time you like

lack commented:
And I'll hold of merging until you let me know that you think it's ready!

ndim committed:
XDG desktop categories do not contain "Multimedia"

When running `desktop-file-validate` on the `*.desktop` file installed
by `soundcraft_setup --install`, it reported that "Multimedia" is not
a valid category according to the Specs:

https://specifications.freedesktop.org/menu-spec/latest/apa.html
https://specifications.freedesktop.org/menu-spec/latest/apas02.html

Therefore, this replaces the "Multimedia" part with "X-Multimedia".

ndim committed:
Update version to 0.4.90 for session-dbus branch

Make it obvious from the version number that this is the partial
rewrite for the session DBus, not the old 0.4.x code.

See https://github.com/lack/soundcraft-utils/pull/22

ndim committed:
contrib: Use busctl variable for busctl command

This allows us to change that command in a single place if required.

ndim committed:
contrib: Define DBus interface using environment variables

Defining the DBus interface strings in a single place prevents
typos and simplifies later changes.

ndim committed:
contrib: Command line DBus example set MASTER_L_R

In the command line DBus example, set the USB audio routing to
MASTER_L_R as that is be available on all three Notepad devices.

Now the example script should run out of the box with any
Notepad model, not just on the Notepad-12FX which the script's
author happens to own.

ndim committed:
dbus.py: Turn it from script into pure library module

Running dbus.py directly is not supported as documented by `serviceExePath()`,
so the `if __name__ == '__main__:'` serves no function.

Directly running dbus.py now is a no-op.

ndim committed:
CONTRIBUTORS.md: Add myself

Add myself to CONTRIBUTORS.md as this is the start of a significant
amount of development effort on my part.

As my main goal here is to make `soundcraft-utils` play well together
with the different parts of different Linux distributions, I am
calling my contribution topic "system integration".

ndim commented on setup.py:
```diff
@@ -32,7 +32,8 @@
         "console_scripts": [
             "soundcraft_ctl=soundcraft.cli:main",
-            "soundcraft_dbus_service=soundcraft.dbus:main",
+            "soundcraft_session_service=soundcraft.dbus:service_main",
```
We could also call the executable `soundcraft_service` instead of `soundcraft_session_service`. I just want to avoid the old name `soundcraft_dbus_service` so that I can explicitly break backwards compatibility with the system bus service.

ndim commented on soundcraft/setup.py:
```diff
@@ -61,22 +61,30 @@ def serviceExePath():
 
 
-def find_datadir():
+def find_datadir(destdir):
```
I probably will change the name from `destdir` to `chroot`.

ndim committed:
setup: Consistent messages for D-Bus and XDG

Consistently use the exact spelling "D-Bus" and "XDG" in the
output messages.

ndim committed:
Change all constant `f""` strings to `""` strings

Change all constant `f""` strings (i.e. those not containing
any `{}` characters) to normal `""` strings.

This makes it clear that at least those strings cannot cause
exceptions due to unknown variables. (`f""` strings are
interpreted at runtime, so tools like `flake8` will not warn
in advance that something inside a f string `{}` will fail.)

ndim committed:
setup: Wait for bus activation after writing service file

As the D-Bus needs to notice a newly created `*.service` file
before it can bus activate the service defined by the file, we
need to give the bus a bit of time for that.

We achieve that by checking every second for a maximum of five
seconds. This drastically reduces the probability of the bus
activation failing from almost certainly to practically never.

The five second timeout ensures failure within a reasonable
time frame for testing.

However, the race condition between the bus noticing the new
`*.service` file and us connecting to the bus activated
service still exists in principle.

To achieve such fine grained control over the service startup,
we are switching from the `soundcraft.dbus.Client` methods to
using the `org.freedesktop.DBus` D-Bus API.  That API also
gives us the opportunity to determine whether the service is
actually running before trying to terminate it so that we only
try to shut it down if it is actually running.

This also removes the dependency of `soundcraft.setup` on the
`soundcraft.dbus.Client` class.  That is quite nice because
the Client 'shutdown()' method always backtraces and pulles in
more dependencies.

ndim committed:
Prepare for installing non-256px sized PNG icon files

The shipped PNG icon file is a 256x256 icon, so we prepare
for adding differently sized PNG icon files and for properly
installing and uninstalling them.

ndim commented:
A quick update on the state of this `session-dbus` branch: I am making progress, but am not quite finished yet.

All the things which need to be fixed before merging are marked in source code comments with `FIXME05` which can be easily found with `git grep -E -i 'fixme05'`.

I still need to

  * do a few more rebasings to rearrange the commits for easier understanding of the changes
  * fix a few more bugs
  * improve the commit messages (grammar, wording, verbosity)
  * re-read our discussions to make sure I have not missed implementing or addressing any of your or my points

Then the next tasks are to

  * discuss a few topics with @lack as to which direction to go for some special cases
  * write a new chapter for `README.md` on installation and a `PACKAGING.md` file to help distro packagers

Then you can go review the entire patch series and you can go over what the commits do and why and what you want changed. Then I can address any issues raised during the review, and then FINALLY @lack can merge and close this.

Things like adding properly written and halfway complete man pages or fixing the annoying exception tracebacks during `client.shutdown()` can wait until after the merge, or even after `0.5.0`.

Also after the merge, I will prepare proper Fedora packages for `soundcraft-utils` (I have a working prototype package built  from my `session-dbus` branch). Packages for other distributions should then also be relatively easy to make with the changed architecture, the new `soundcraft_setup` tool and the `PACKAGING.md` documentation. (I am talking packages which work out of the box, as opposed to the packages currently linked from `README.md` which appear to me to completely avoid the dbus and xdg installation, and thus also the issue of manually needing to install the current running-as-root system bus service.)

FTR, according to `git log | grep ^Author: | sort | uniq -c | sort -bgr` there are 61 commits by @ndim and 81 commits by @lack, but I do expect the big `session-dbus` merge to happen definitely before I have caught up on that number. :-)

ndim committed:
contrib: Add commit-author-stats.sh (provide commit author overview)

Provide an overview of the commit authors of the current tree's history.

Example output could look like

     42 Author: Jane Doe <jdoe@example.com>
     23 Author: Ian Diot <idiot@example.com>
      1 Author: Moo Cow <moocow@example.com>

Note: The script contains a special case for one of Jim's early commits.

ndim committed:
contrib: add sloccount stats

Let David A. Wheeler's 'SLOCCount' count the source code lines
and apply metrics regarding effort and development costs etc.

Partial example output (as generated using David A. Wheeler's
'SLOCCount'):

```
SLOC	Directory	SLOC-by-Language (Sorted)
877     soundcraft      python=877
163     tools           python=100,sh=63
77      test            python=77
20      contrib         sh=20


Totals grouped by language (dominant language first):
python:        1054 (92.70%)
sh:              83 (7.30%)
```

ndim committed:
contrib: Add todo.sh to list TODO and FIXME from the sources

Add one script to find all TODOs and FIXMEs in the source tree.

This has a bit of an issue with multi-line TODOs and FIXMEs,
but it is a good start.

Example output:

```
[user@host soundcraft-utils]$ ./contrib/sources/todo.sh
README.md:TODO
soundcraft/gui.py:            # Todo: Can we relaunch ourselves?
soundcraft/notepad.py:        # TODO: Decode these?
[user@host soundcraft-utils]$
```

ndim committed:
Improve gi module import error message

Improve gi module import error message in the `soundcraft.dbus`
module.

Some distributions name their python packages `python-*`,
others name them `python3-*`. Mention both variants here.

Also print a trailing newline, so that the printed message
is framed by an empty line each on the top and on the bottom.

Use the same way of importing `gi` in the `soundcraft.gui`
module, which did not have any messages before regarding
packages to install in order to provide the `gi` module.

ndim committed:
In messages, use "D-Bus" spelling consistently

In program messages and documentation, consistently use
the spelling "D-Bus" instead of unnecessarily mixing it
up with "DBUS" and "dbus".

ndim committed:
Start collecting constants in module: executable names

Start collecting constants in a dedicated python module
`soundcraft.constants` by defining the basenames of all
executables there and using them everywhere.

The standalone `soundcraft.constants` module not importing
any other modules makes it possible to e.g. import this module
from the setuptools `setup.py` script in the top level directory
without drawing in any complex build time dependencies to other
modules.

Note that the actual location of the source file imported as
`soundcraft.constants` could differ depending on what the user
has set `PYTHONPATH` to or modified `sys.path` to in some startup
file.  Therefore, we verify that we have actually imported
the `soundcraft/constants.py` source file which is located below
the directory above the `setup.py`.

Importing the constants module both avoids the difficulty of
regexp parsing a programming language of which only a subset can
be parsed with regexps, and also allows us to use more of the
power of a programming language in the `soundcraft.constants`
module like using some constant values to compose others.

ndim committed:
Move (un)install code into `soundcraft_setup` script

Moves code from the `soundcraft.dbus` module which is not
related to running the D-Bus client or the D-Bus service to
a new module `soundcraft.setup`.  That module then implements
the new script `soundcraft_setup`.

This split in turn allows us to remove the setup code (script,
module, and data files) after use if we do not need it any more,
e.g. when preparing a binary package for a Linux distribution.

ndim committed:
README.md: Mark sections needing a rewrite for session D-Bus service

Mark the `README.md` sections needing a rewrite for the session
D-Bus service with `FIXME05` (i.e. to be fixed before release 0.5).

ndim committed:
HACKING.md: Clarify meaning

ndim reviewed (commented):
I am undecided whether allowing multiple GUI app instances simultanously with `NON_UNIQUE` should be a debug-only commit to be removed from the `session-dbus` branch before merging, or whether this should eventually be merged and made permanent.

@lack Do you have an opinion on `NON_UNIQUE`?

ndim committed:
HACKING.md: Things to avoid for compat with Python >= 3.6

List some things to avoid for compatibility with Python >= 3.6
which I had accidentally wanted to use.

ndim committed:
contrib: check for mistakes in code

Look for some easy to find and easy to miss mistakes in
the code by running a few searches for regexp patterns of
mistakes with `git grep`:

  * The `functools.cached_property` decorator was only added
    in Python 3.8, but we aim for compatiblity with Python 3.6.

  * The `missing_ok=` parameter to `Path().unlink()` was only
    added in Python 3.8, but we aim for compatibility with 3.6.

Example output if no mistake patterns have been found:

```
[user@host soundcraft-utils]$ ./contrib/sources/find-mistakes.bash
[user@host soundcraft-utils]$ echo $?
0
[user@host soundcraft-utils]$
```

There will be a non-0 exit code if any mistake has been found.

ndim committed:
Use absolute imports everywhere

Consistently use absolute imports everywhere in the codebase
by adapting `soundcraft.cli` to follow the import pattern
used by all the other source files.

Closes: https://github.com/lack/soundcraft-utils/issues/25

ndim committed:
Replace 'soundcraft-utils' strings by const.PACKAGE

Replace the literal `'soundcraft-utils'` string by references to
`const.PACKAGE` to prevent inconsistencies and typos.

From a consistency point of view, this would be a good point to
replace the existing `__version__` definition and the regexp
parsing of that expression by just defining `const.VERSION`.

ndim committed:
gui: allow multiple instances of the GUI app

Allow multiple instances of the GUI app to run at the same time.

This can be useful to

   * test the communication via the D-Bus service by having two
     GUI windows side by side (for developers)

   * have a GUI window on more than one virtual desktop (for users,
     like other audio settings programs like pavucontrol)

However, some programs have started to be single GUI window only
which forces virtual desktop switches, but that might be a new
Gnome desktop policy other just changed default behaviour in Gtk
without much more background.

Note that `flake8` and `black` disagree about how the expression
ORing the binary constants is supposed to be formatted, so we
resolve the conflict by telling `flake8` to ignore what `black`
produces.

ndim committed:
Replace Gtk App ID and XDG Icons by const.CONSTANT

Collect string constants for the Gtk Application ID and the XDG
Icon names and put them into the `soundcraft.constants` module
to allow changing them later in an single place.

ndim committed:
setup: Consistently name install functions

Consistently name the functions installing files something with
"install" in them instead of "setup".

lack committed:
Merge pull request #23 from ndim/cleanups-before-session-dbus

Cleanups before `session-dbus` branch

lack committed:
Add fedora-specific installation prerequisite steps

ndim committed:
tools/link_system_libs: Made it idempotent

Repeated runs of `tools/link_system_libs` used to produce
error messages like

    ln: failed to create symbolic link '/home/user/.local/share/virtualenvs/soundcraft-utils-ABCDEFG/lib/python3.8/site-packages/gi': File exists

Now the script checks if the symlink already exists and
points to the correct place, and will accept that as a
successful script run as well.

If the symlink already exists and points to the wrong place,
an error message will still be produced.

ndim committed:
setup: Create and use common dir detection functions

Move the `datadir` detection into a separate function so we
can reuse the code and use consistent directories from multiple
places in the source code.

This should detect `datadir` correctly for installs into
`/usr/local`, `/usr`, `~/.local` and
`~/.local/share/virtualenvs/soundcraft-utils-ABCDEFG`.

When running from a virtualenv located in `$HOME`, install the
D-Bus and XDG config files to `$HOME/.local` so the configuration
is actually found by the D-Bus and can actually take effect (neither
D-Bus nor XDG take a look anywhere into `~/.local/share/virtualenvs/`).

As `dbus1_root` (formerly known as `cfgroot`) is now defined in
terms of `datadir`, it should install alongside the respective
XDG config locations which are where D-Bus is looking for its
configuration files.

Remove the unused `cfgroot` parameter to the `install_dbus()`
and `uninstall_dbus()` functions to make things simpler.

lack committed:
Merge pull request #27 from ndim/fix/idempotent-link_system_libs-script

tools/link_system_libs: Made it idempotent

lack commented on soundcraft/gui.py:
```diff
@@ -237,9 +257,41 @@ def close_cb(self, action, parameter):
+            flags=(
+                0
+                | Gio.ApplicationFlags.HANDLES_COMMAND_LINE  # noqa: W503
+                | Gio.ApplicationFlags.NON_UNIQUE  # noqa: W503
```
I don't feel to strongly about it either way... My understanding of this flag is that it would prevent more than one gui from running per session bus; which would effectively mean a 1:1 gui-to-service mapping.  That seems reasonable to do, but not really necessary either.

Do you have a good use-case in mind for why we would want `NON_UNIQUE` set?

ndim replied:
I can think of two use cases for which I might want to use more than one GUI window:

  * Use two GUI windows side by side to make sure both the GUI and the D-Bus service work as intended, when I click on "Apply" in one window, and the other window updates instantly. Mainly interesting for developers.

  * Use one GUI window on one virtual desktop, and another GUI window on another virtual desktop. Depending on how I have arranged my applications on my set of virtual desktops, this might be useful. Another widely used tool for audio settings, namely `pavucontrol` allows me to do just that. Interesting for users.

As `soundcraft_gui` is significantly less resource intensive than something like, say, libreoffice, having multiple instances running at the same time does not appear to be a resource issue either.

ndim committed:
HACKING.md. Add link to Desktop Menu Specification

ndim commented:
FYI, I am trying to get a first set of about 20 commits in a stable state so you can review them, while I still need to work on the second set of about 20 commits finishing up what the first set has started.

ndim committed:
setup: Improve D-Bus message output

More accurate and more verbose output from `soundcraft_setup` during
D-Bus service test startup.

ndim committed:
setup: After D-Bus service test start, also shut it down

After the D-Bus service test start during `soundcraft_setup --install`,
also shut the service down as well.

This both verifies the shutdown procedure works, and forces normal bus
activation of the service on the next D-Bus client startup.

Also, this means `soundcraft_setup` does not leave any processes
running which is a cleaner thing to do.

ndim committed:
Generate, install and uninstall PNG icons of different sizes

It used to be that `soundcraft_dbus_service --setup` installed
the same 256x256 PNG file as the PNG icon for any number of sizes
which were not 256x256 at all.

We now have all the actual sizes in the git repo so we both
install and uninstall them: The old sizes we installed the wrong
sized icon to (16, 24, 32, 48), a few new sizes which appear to
be in popular use (64, 96, 128, 192), as well as the old 256.

We achieve this by converting the source `*.svg` icon to a number
of `*.png` icon files in different sizes calling `inkscape` from
a script, and then adding those different PNG files to the git tree
alongside the original 256x256 PNG file.  Adding the PNG files to
the git repository avoids a requirement for having `inkscape´
available when installing from source.

All the PNG files each are smaller in bytes than the SVG file is,
and we do not expect (m)any changes to the icon over time, so the
churn in binary files should be acceptable even though we now have
nine PNG files instead of just one.

As icon changes are expected to be very rare, we do not really need
a pre-commit check to run `tools/update-icons.sh`.

This all only makes sense if there are many commonly used desktop
environment installations which cannot properly use icons in the
scalable SVG format and work better when desktop icons and Gtk App
window icons are in PNG format.

ndim committed:
setup: Reindent code preparing for the next commit

This should help read the commit diffs.

ndim committed:
setup: Reorder the code, preparing for the next commit

This should make the next commit diff easier to read.

ndim committed:
setup: Restructure into classes, directly install XDG files

Restructure the installation parts into classes: One class for
installing/uninstalling the D-Bus service, and one class for
installing/uninstalling the XDG desktop file and icons.

Introduce templating to the `*.desktop` file to install it with
the correct Gtk app id and a full path to the `soundcraft_gui`
executable.

Stop running `xdg-*` programs for installing XDG desktop and icon
files and directly install the files in the appropriate directory
instead.

Also, stop installing the 256px sized PNG icon file as if it were
of size 16, 24, 32, and 48.

Note: I am not sure whether directly installing the XDG files is
the correct way, though. It might be that `soundcraft_setup` is
supposed to call some `xdg-*` program after all. It might also be
that whoever is calling `soundcraft_setup` (a user or a packager)
is supposed to call some `xdg-*` program after `soundcraft_setup`
has finished. We will have to find that out and change things
later, if necessary.

ndim committed:
Move the D-Bus service to the session bus

Move the D-Bus service from the system bus to the session bus:

  * Change the D-Bus service and client code to use
    `SessionBus()` instead of `SystemBus()`.

  * Explicitly break compatibility with the old D-Bus service
    by changing the dbus service executable name from
    `soundcraft_dbus_service` to `soundcraft_session_service`,
    and the entrypoint function from `main()` to `service_main()`.

  * Adapt the D-Bus service configuration: Remove `User=root`,
    move the `*.service` file, and stop installing the system
    service policy file).

  * Change program messages regarding the bus used, `--setup`,
    and the removal of the need to run as root.

  * Change the `dbus-send` and `busctl` example commands in
    `contrib/` to use the session bus.

ndim committed:
setup: Use same src file iteration

Use the same iteration loop over the source files for the D-Bus
files as for the XDG files: Whenever a certain source file comes
up, handle that source file according to its specific needs.

For the `*.service` file, that means a standard string template
substitution and writing the content to a location defined by code.

Other files are just copied over as they are.

Once we use the same loop in multiple places, we can refactor the
code to extract the common code into a separate function.

ndim reviewed (commented):
This is the first batch of 25 commits for session-bus which you can review, and after I have fixed the issues found in the review and force-pushed a new sequence of commits, merge into [lack:session-dbus](https://github.com/lack/soundcraft-utils/tree/session-dbus).

In the mean time, I will try to finish the second batch of about 30 commits.

ndim reviewed (commented):
[message deleted due to wrong use of github.com UI]

ndim commented on contrib/sources/sloccount-stats.sh:
```diff
@@ -0,0 +1,5 @@
+
+cd "$(dirname $(dirname $(dirname "$0")))"
+
+sloccount soundcraft contrib test tools
```
I just found this interesting. This is not required for session-dbus at all.

Feel free to pick this up or drop it.

ndim commented on contrib/sources/commit-author-stats.sh:
```diff
@@ -0,0 +1,7 @@
+
+git log | grep ^Author: \
+    | sed 's|^Author: lack <i.am@jimramsay.com>$|Author: Jim Ramsay <i.am@jimramsay.com>|' \
+    | sort | uniq -c | sort -bgr
```
I just found this interesting. This is not required for session-dbus at all.

Feel free to pick this up or drop it.

ndim commented on contrib/sources/todo.sh:
```diff
@@ -0,0 +1,6 @@
+cd "$(dirname $(dirname $(dirname "$0")))"
+
+# Using [] character class to prevent this line from finding itself.
+git grep -i -E '([f]ixme|[t]odo)'
```
This can be useful for developing in general, and for keeping track of the miscellaneous FIXME locations in the session-dbus branch in particular.

I would prefer if that could be picked up.

ndim commented on contrib/sources/find-mistakes.bash:
```diff
@@ -0,0 +1,27 @@
+    find_py_mistake "$mistake_pattern"
+done
+
+exit 0
```
This has helped me get rid of some Python-3.8-isms in the code while working on the session-dbus branch. It is not really required any more.

If you like it for the future, pick it up. Otherwise, I am OK with dropping it.

ndim commented on CONTRIBUTORS.md:
```diff
@@ -5,6 +5,7 @@ Contributors
+- [Hans Ulrich Niedermann](mailto:hun@ndimensional.de) - System integration
 
 Artwork
 -------
```
Wait a minute... there is a typo in the mail address. I will fix that in the next rebase.

ndim commented on setup.py:
```diff
@@ -22,7 +22,7 @@
 setup(
-    name="soundcraft-utils",
+    name=const.PACKAGE,
     version=version,
```
If you concur that replacing the `__version__` thing and parsing Python code with a regexp with importing a module of constants for a `const.VERSION` makes sense, I would add such a commit after this one on the next rebase.

ndim commented on setup.py:
```diff
@@ -45,6 +45,7 @@
         "console_scripts": [
             f"{const.BASE_EXE_CLI}=soundcraft.cli:main",
             f"{const.BASE_EXE_SERVICE}=soundcraft.dbus:main",
+            f"{const.BASE_EXE_SETUP}=soundcraft.setup:main",
```
I think `soundcraft_setup` for the executable and `soundcraft.setup` as the module for the code are reasonable names for the install/uninstall utility.

If you prefer something better, I can relatively easily change those names due to `const.BASE_EXE_SETUP` being used.

ndim commented on soundcraft/setup.py:
```diff
@@ -78,10 +78,33 @@ def serviceExePath():
-SCALABLE_ICONDIR = Path("/usr/local/share/icons/hicolor/scalable/apps/")
-
-
-def install_dbus(cfgroot=Path("/usr/share/dbus-1")):
```
This used to be a weird mix of install locations, when both XDG and D-Bus are looking into `~/.local/share`, `/usr/local/share`, and `/usr/share` (in that sequence).

ndim commented on README.md:
```diff
@@ -40,6 +40,8 @@ Installation
 ### Installation
 
+FIXME05 This section needs to be rewritten for the session dbus
+
```
This could also be moved to later in the commit sequence and then be rebased away with the updated version of `README.md`, if you prefer that sequence of commits.

ndim commented on soundcraft/setup.py:
```diff
@@ -109,89 +109,93 @@ def find_datadir():
-
-    print("D-Bus installation is complete")
-    print(f"Run {const.BASE_EXE_GUI} or {const.BASE_EXE_CLI} as a regular user")
+    if True:
```
As the code will be moved into object methods which are indented two steps shortly, this indents the code first.

The advantage to this is that the commit diff actually splitting functions into object methods becomes more readable. The disadvantage is having an additional commit which does not really change the code.

If you do not like this, I can squash this commit together with the following actual commit into one commit, for which the commit diff will be more difficult to read.

ndim commented on soundcraft/setup.py:
```diff
@@ -167,6 +167,31 @@ def install_dbus():
         print(f"Run {const.BASE_EXE_GUI} or {const.BASE_EXE_CLI} as a regular user")
 
 
+def uninstall_dbus():
```
Reorder the functions, grouping them in the same sequence the next commit will have that code in object methods.

This makes the next commit diff a lot easier to read.

Same as before... if you do not like these extra commits for patch readability, this can be easily squashed into the next commit for less commits, while costing some commit diff readability.

ndim reviewed (commented):
As installing files to well-known locations using Python's setuptools is non-trivial, this means that I need to write a comprehensive set of testcases for **installing** `soundcraft-utils`. This falls outside the usual test frameworks, so I need to roll my own, and the whole thing is ugly and a mess which I have not found the motivation to tackle yet.

ndim commented:
This has been concluded in https://github.com/socratools/socranop/pull/4.
